File: 3154-maximum-value-of-an-ordered-triplet-i/maximum-value-of-an-ordered-triplet-i.java
class Solution {
    public long maximumTripletValue(int[] nums) {
        long res = 0, maxi = 0, maxDiff = 0;
        for (int num: nums) {
            res = Math.max(res, maxDiff * num);
            maxDiff = Math.max(maxDiff, maxi - num);
            maxi = Math.max(maxi, num);
        }
        return res;
    }
}

File: 3154-maximum-value-of-an-ordered-triplet-i/maximum-value-of-an-ordered-triplet-i.py
class Solution:
    def maximumTripletValue(self, nums: List[int]) -> int:
        res = maxi = max_diff = 0
        for num in nums:
            res = max(res, max_diff * num)
            max_diff = max(max_diff, maxi - num)
            maxi = max(maxi, num)
        return res